File: pickr/lib/classes/games/briscola_chiamata.dart
import 'package:pickr/classes/card.dart';
import 'package:pickr/classes/games/game.dart';
import 'package:pickr/classes/player.dart';
import 'package:pickr/enums/suits.dart';
import 'package:pickr/exceptions/out_of_bound_exception.dart';

class BriscolaChiamata extends Game {
  //
  /// Number of beginning cards for each player.
  final int _numplayers = 5;

  /// Number of players.
  ///
  /// DEBUG: this value is set to 4 for debug purposes. Its real values should
  /// be selected by the user in the game options selection page.
  final int _numcards = 8;

  /// Briscola suit selected for the game
  Suit _briscola;

  @override
  void start() {
    //
    deck.init();

    deck.shuffle();

    _briscola = deck.cards.last.suit;

    spread();
  }

  @override
  void addPlayer({String id}) {
    if (players.length <= _numplayers - 1) {
      players.add(Player(id: id));
    } else {
      throw OutOfBoundException();
    }
  }

  @override
  void spread() {
    players.forEach((player) {
      player.hand = deck.picks(_numcards);
    });
  }

  @override
  void gameround() {
    players.forEach((player) => dropCard(card: player.drop()));

    var winningCard = rule();

    winner(winningCard);

    table.clear();
  }

  @override
  GamingCard rule() {
    var winningCard = table.first;

    table.forEach((card) {
      if (card.val > winningCard.val) {
        // T T T
        if (winningCard.suit == _briscola && card.suit == _briscola) {
          winningCard = card;
        }
        // T F F
        if (winningCard.suit != _briscola && card.suit != _briscola) {
          winningCard = card;
        }
      }
      // F T F
      else if (card.suit == _briscola && winningCard.suit != _briscola) {
        winningCard = card;
      }
    });

    return winningCard;
  }
}

File: pickr-dbc/src/pickr/games/BriscolaChiamata.java
package pickr.games;

import pickr.classes.GamingCard;
import pickr.classes.Player;
import pickr.enums.Suit;
import pickr.exceptions.OutOfBoundException;
import pickr.exceptions.WrongCardNumberException;

public class BriscolaChiamata extends Game {
	
	int numcards = 8;
	
	private Suit briscola;

	public BriscolaChiamata(int numplayers, int score) {
		super(numplayers, score);
	}

	@Override
	void spread() throws OutOfBoundException {
		for(Player player : players) {
			player.hand(deck.picks(numcards));
		}		
		assert(this.deck.getCards().size() == 0);
	}

	@Override
	void start() throws WrongCardNumberException, OutOfBoundException {
		this.deck.init();
		briscola = deck.getCards().get(deck.getCards().size()-1).getSuit();
		this.spread();
	}

	@Override
	GamingCard rule() {
		
		GamingCard winningCard = this.getTable().get(0);
		
		for(GamingCard card : this.getTable()) {			
			
			if(card.getVal() > winningCard.getVal()) {
				if(winningCard.getSuit() == briscola && card.getSuit() == briscola) {
					winningCard = card;
				}
				if(winningCard.getSuit() != briscola && card.getSuit() != briscola) {
					winningCard = card;
				}
			}
			else if(card.getSuit() == briscola && winningCard.getSuit() != briscola) {
				winningCard = card;
			}			
		}
				
		return winningCard;
	}
}
